### DISCUSSION TITLE
OpenAPI definition in SwaggerUI is almost empty

### DISCUSSION THREAD

**Discussion by clairecassan** [PostgREST/postgrest#1877] [Q&A] [answered]
Hi all,
I have been using postgrest for a few months, successfully. I am displaying the OpenAPI definition in Swagger and it was working well until I pointed the server on an upgraded schema (new tables have been added). I can still query data with REST calls, but the Swagger page is almost empty, I only have the "Introspection" entry. I was expecting to see the same entries as before, plus the get/post/delete for the new tables
The only thing I changed on my side was the conf file (db-uri,db-schema,db-anon-role). The postrest server and the React code is unchanged. 
Does anyone know what could impact the OpenAPI definition file and make it so that none of the tables are visible? 
Thank you
CC

**Comment by laurenceisla** [comment, marked as answer]
PostgREST returns the OpenApi definition for the anonymous role, maybe this new anon role you created doesn't have any permissions on the database objects.

Edit:
It's also possible to get the OpenApi definition for an authenticated user if you send the JWT through the Authorization header.

**Reply by clairecassan** [reply]
Thank you for your quick answer. I checked your suggestion on the anon role. I asked the admin to add an upgraded schema on the original db I was using, so now I have the same db-uri and db-anon-role as before, and the problem persists. Is there a way to debug the generation of the OpenApi definition? 
Our schemas are complex, around 240 tables and 4000 columns, but we only added 4 new tables so maybe I can debug to see what's wrong with these tables

**Reply by laurenceisla** [reply]
The things that come to mind are:
- Maybe the [schema cache needs reloading](https://postgrest.org/en/stable/admin.html#schema-reloading). It should be done every time a table, view or function is created on the database to refresh the OpenApi definition.
- If you have more than one schema, use the [`Accept-Profile` header](https://postgrest.org/en/stable/api.html?highlight=openapi#switching-schemas) for the schema you want to access.

Don't know of a way to debug the OpenApi generation, though.

**Reply by clairecassan** [reply]
In my tests I'm recreating the docker image and restarting the container that hosts the postgrest server each time, so the cache should be ok. Switching schema at run-time works well, I have been using this feature for a while and I can switch between old and new schema to trigger a get on a table. The only thing that seems broken is the OpenAPI definition file. I will continue my tests today, and make sure I can run get calls on the new tables, so far I have only been running get calls on tables common to the 2 schemas.

**Reply by clairecassan** [reply]
I was able to run get calls on the new tables. And after a clean/start of the whole system (backend/frontend) the OpenAPI file is now properly generated. I'm not sure what was wrong, maybe something on the nginx layer, or a network issue.
Thank you for having taken the time to answer my question.

**Comment by dnk8n** [comment]
> Edit:
> It's also possible to get the OpenApi definition for an authenticated user if you send the JWT through the Authorization header.

How can one set the Authorization header for swagger UI so that it displays the interactive protected endpoints? Can it be set dynamically from the client?

**Reply by laurenceisla** [reply]
To show all the protected endpoints you have these options: 
- Request the OpenAPI endpoint (by default at root `/`) sending the JWT in the Authorization header `Authorization: Bearer <the_jwt>`.
- Show all the endpoints regardless of permissions by setting: [`openapi-mode="ignore-privileges"`](https://postgrest.org/en/latest/configuration.html#openapi-mode). 

If you need to add the security definitions, that is, to get the "Authorize" button in the Swagger UI, you need to set this config variable: [`openapi-security-active=true`](https://postgrest.org/en/latest/configuration.html#openapi-security-active).

Edit: Ah, nvm, I misunderstood. I just saw the new discussion you opened here:  https://github.com/PostgREST/postgrest/discussions/2716
